cocos/2d/renderer/batcher-2d.ts

```diff
@@ -794,41 +794,44 @@ export class Batcher2D implements IBatcher {
         // TODO Set opacity to ui property's opacity before remove it
         // @ts-expect-error temporary force set, will be removed with ui property's opacity
         uiProps._opacity = opacity;
-        if (uiProps.colorDirty) {
+        if (opacity > 0) {
+            if (uiProps.colorDirty) {
             // Cascade color dirty state
-            this._opacityDirty++;
-        }
+                this._opacityDirty++;
+            }
 
-        // Render assembler update logic
-        if (render && render.enabledInHierarchy) {
-            render.fillBuffers(this);// for rendering
-        }
+            // Render assembler update logic
+            if (render && render.enabledInHierarchy) {
+                render.fillBuffers(this);// for rendering
+            }
 
-        // Update cascaded opacity to vertex buffer
-        if (this._opacityDirty && render && !render.useVertexOpacity && render.renderData && render.renderData.vertexCount > 0) {
+            // Update cascaded opacity to vertex buffer
+            if (this._opacityDirty && render && !render.useVertexOpacity && render.renderData && render.renderData.vertexCount > 0) {
             // HARD COUPLING
-            updateOpacity(render.renderData, opacity);
-            const buffer = render.renderData.getMeshBuffer();
-            if (buffer) {
-                buffer.setDirty();
+                updateOpacity(render.renderData, opacity);
+                const buffer = render.renderData.getMeshBuffer();
+                if (buffer) {
+                    buffer.setDirty();
+                }
```

Comment:
If opacity turn to zero in the current frame, but the render data will potentially contain un-updated data, make sure you have decoupled data update and rendering very carefully

Reply:
If the opacity of the current frame becomes 0, then the content in renderData will not be rendered and submitted, and there is no point in updating the data in renderData at this point. The component's colorDirty will not be reset until its opacity becomes non-zero, and then the renderData update logic will be performed.